policy-controller/k8s/index/src/outbound_index.rs

```diff
@@ -0,0 +1,361 @@
+        let routes = ns.service_routes_or_default(key);
+        Ok(routes.watch.subscribe())
+    }
+
+    pub fn lookup_service(&self, addr: IpAddr) -> Option<ServiceRef> {
+        self.services.get(&addr).cloned()
+    }
+}
+
+impl Namespace {
+    fn apply(&mut self, route: HttpRoute) {
+        tracing::debug!(?route);
+        let name = route.name_unchecked();
+        let outbound_route = match self.convert_route(route.clone()) {
+            Ok(route) => route,
+            Err(error) => {
+                tracing::error!(%error, "failed to convert HttpRoute");
+                return;
+            }
+        };
+        tracing::debug!(?outbound_route);
+
+        for parent_ref in route.spec.inner.parent_refs.iter().flatten() {
+            if parent_ref.kind.as_deref() == Some("Service") {
+                if let Some(port) = parent_ref.port {
+                    if let Some(port) = NonZeroU16::new(port) {
+                        let service_port = ServicePort {
+                            port,
+                            service: parent_ref.name.clone(),
+                        };
+                        tracing::debug!(
+                            ?service_port,
+                            route = route.name_unchecked(),
+                            "inserting route for service"
+                        );
+                        let service_routes = self.service_routes_or_default(service_port);
+                        service_routes.apply(name.clone(), outbound_route.clone());
+                    } else {
+                        tracing::warn!(?parent_ref, "ignoring parent_ref with port 0");
+                    }
+                } else {
+                    tracing::warn!(?parent_ref, "ignoring parent_ref without port");
+                }
+            } else {
+                tracing::warn!(
+                    parent_kind = parent_ref.kind.as_deref(),
+                    "ignoring parent_ref with non-Service kind"
+                );
+            }
+        }
+    }
+
+    fn update_service(&mut self, name: String, service: ServiceInfo) {
+        tracing::debug!(?name, ?service, "updating service");
+        for (svc_port, svc_routes) in self.service_routes.iter_mut() {
+            let opaque = service.opaque_ports.contains(&svc_port.port);
+            svc_routes.set_opaque(opaque);
+        }
+        self.services.insert(name, service);
+    }
+
+    fn delete(&mut self, name: String) {
+        for service in self.service_routes.values_mut() {
+            service.delete(&name);
+        }
+    }
+
+    fn service_routes_or_default(&mut self, service_port: ServicePort) -> &mut ServiceRoutes {
+        let authority = format!(
+            "{}.{}.svc.{}:{}",
+            service_port.service, self.namespace, self.cluster_domain, service_port.port
+        );
+        self.service_routes
+            .entry(service_port.clone())
+            .or_insert_with(|| {
```

Comment:
nit, not a big deal: we only need to format the authority if we are inserting a new `ServicePort`...could this be moved into the `or_insert_with` closure?
```suggestion
        self.service_routes
            .entry(service_port.clone())
            .or_insert_with(|| {
                    let authority = format!(
                        "{}.{}.svc.{}:{}",
                        service_port.service, self.namespace, self.cluster_domain, service_port.port
                    );
```